core/modules/random_split.cc

```diff
@@ -0,0 +1,119 @@
+// contributors may be used to endorse or promote products derived from this
+// software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "random_split.h"
+#include <string>
+#include <time.h>
+
+static inline bool is_valid_gate(gate_idx_t gate) {
+  return (gate < MAX_GATES || gate == DROP_GATE);
+}
+
+const Commands RandomSplit::cmds = {
+    {"set_droprate", "RandomSplitCommandSetDroprateArg",
+     MODULE_CMD_FUNC(&RandomSplit::CommandSetDroprate), Command::THREAD_UNSAFE},
+    {"set_gates", "RandomSplitCommandSetGatesArg",
+     MODULE_CMD_FUNC(&RandomSplit::CommandSetGates), Command::THREAD_UNSAFE}};
+
+CommandResponse RandomSplit::Init(const bess::pb::RandomSplitArg &arg) {
+  double drop_rate = arg.drop_rate();
+  if (drop_rate < 0 || drop_rate > 1) {
+    return CommandFailure(EINVAL, "drop rate needs to be between [0, 1]");
+  }
+  threshold_ = drop_rate * kRange;
+
+  if (arg.gates_size() > MAX_SPLIT_GATES) {
+    return CommandFailure(EINVAL, "no more than %d gates", MAX_SPLIT_GATES);
+  }
+  ngates_ = arg.gates_size();
+
+  for (int i = 0; i < arg.gates_size(); i++) {
+    int elem = arg.gates(i);
+    gates_[i] = elem;
+
+    if (!is_valid_gate(gates_[i])) {
+      return CommandFailure(EINVAL, "Invalid gate %d", gates_[i]);
+    }
+  }
+
+  return CommandSuccess();
+}
+
+CommandResponse RandomSplit::CommandSetDroprate(
+    const bess::pb::RandomSplitCommandSetDroprateArg &arg) {
+  double drop_rate = arg.drop_rate();
+  if (drop_rate < 0 || drop_rate > 1) {
+    return CommandFailure(EINVAL, "drop rate needs to be between [0, 1]");
+  }
+  threshold_ = drop_rate * kRange;
+
+  return CommandSuccess();
+}
+
+CommandResponse RandomSplit::CommandSetGates(
+    const bess::pb::RandomSplitCommandSetGatesArg &arg) {
+  if (arg.gates_size() > MAX_SPLIT_GATES) {
+    return CommandFailure(EINVAL, "no more than %d gates", MAX_SPLIT_GATES);
+  }
+
+  for (int i = 0; i < arg.gates_size(); i++) {
+    int elem = arg.gates(i);
+    gates_[i] = elem;
+
+    if (!is_valid_gate(gates_[i])) {
+      return CommandFailure(EINVAL, "Invalid gate %d", gates_[i]);
+    }
+  }
+  ngates_ = arg.gates_size();
```

Comment:
We should probably verify _all_ the `arg.gates(i)` values before setting any of the `gates_[i]` values, but in any case, it would certainly be better to verify `elem` before assigning it to `gates_[i]` here, so that we don't leave a bogus value in the `gates_` array.

It would also be nice to have both `::Init` and `::CommandSetGates` use the same verification and setting code (though that's less important—anything bogus during `Init` gets wiped away, as if it never happened...).